src/main/java/com/uid2/admin/vertx/service/ClientSideKeypairService.java

```diff
@@ -0,0 +1,238 @@
+package com.uid2.admin.vertx.service;
+
+import com.uid2.admin.secret.IKeypairGenerator;
+import com.uid2.admin.secret.IKeypairManager;
+import com.uid2.admin.store.Clock;
+import com.uid2.admin.store.reader.RotatingSiteStore;
+import com.uid2.admin.store.writer.ClientSideKeypairStoreWriter;
+import com.uid2.admin.util.InputUtil;
+import com.uid2.admin.vertx.ResponseUtil;
+import com.uid2.admin.vertx.WriteLock;
+import com.uid2.shared.auth.Role;
+import com.uid2.shared.middleware.AuthMiddleware;
+import com.uid2.shared.model.ClientSideKeypair;
+import com.uid2.shared.store.reader.RotatingClientSideKeypairStore;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.KeyPair;
+import java.time.Instant;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static com.uid2.admin.store.writer.ClientSideKeypairStoreWriter.toJsonWithPrivateKey;
+import static com.uid2.admin.store.writer.ClientSideKeypairStoreWriter.toJsonWithoutPrivateKey;
+
+public class ClientSideKeypairService implements IService, IKeypairManager {
+    private final AuthMiddleware auth;
+    private final Clock clock;
+    private final WriteLock writeLock;
+    private final ClientSideKeypairStoreWriter storeWriter;
+    private final RotatingClientSideKeypairStore keypairStore;
+    private final RotatingSiteStore siteProvider;
+    private final IKeypairGenerator keypairGenerator;
+    private final String publicKeyPrefix;
+    private final String privateKeyPrefix;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClientSideKeypairService.class);
+
+    public ClientSideKeypairService(JsonObject config,
+                                    AuthMiddleware auth,
+                                    WriteLock writeLock,
+                                    ClientSideKeypairStoreWriter storeWriter,
+                                    RotatingClientSideKeypairStore keypairStore,
+                                    RotatingSiteStore siteProvider,
+                                    IKeypairGenerator keypairGenerator,
+                                    Clock clock) {
+        this.auth = auth;
+        this.writeLock = writeLock;
+        this.storeWriter = storeWriter;
+        this.keypairStore = keypairStore;
+        this.keypairGenerator = keypairGenerator;
+        this.siteProvider = siteProvider;
+        this.clock = clock;
+        this.publicKeyPrefix = config.getString("client_side_keypair_public_prefix");
+        this.privateKeyPrefix = config.getString("client_side_keypair_private_prefix");
```

Comment:
it's a bit not nice that there's a prefix for every key type but the IdentityScope in the prefix should be set once and reused in all of the key type prefixes. I guess room for improvement in the future.